analytics/lib/data_collectors.py

```diff
@@ -0,0 +1,247 @@
+from django.db import connection
+from analytics.models import DeploymentCount, RealmCount, UserCount, StreamCount, PMCount, HuddleCount
+from zerver.models import Realm, UserProfile, Message, Stream
+from datetime import timedelta
+
+
+# add sum smallest_interval
+class AnalyticsStat:
+    def __init__(self, property, zerver_table, filter_args, analytics_table, smallest_interval, frequency):
+        self.property = property
+        self.zerver_table = zerver_table
+        # might have to do something different for bitfields
+        self.filter_args = filter_args
+        self.analytics_table = analytics_table
+        self.smallest_interval = smallest_interval
+        self.frequency = frequency
+
+COUNT_TYPES = frozenset(['zerver_pull', 'time_aggregate', 'cross_table_aggregate'])
+
+
+# todo: incorporate these two arguments into a universal do_pull function.
+
+
+def count_query(stat):
```

Comment:
I realized as I was writing the doc: we should just have a ZerverPull class (or some better name), with fields for filtered_zerver_table, analytics_table, and query. Then AnalyticsStat can just have a ZerverPull instance, and we can get rid of AnalyticsStat.analytics_table and AnalyticsStat.zerver_table. I think this will make more sense given that many of our queries require joins on the zerver side.